### ISSUE TITLE
Avoid sync fs-calls in backupMainService

### ISSUE THREAD

**Issue by jrieken** [microsoft/vscode#16375]
The `backupMainService` runs synchronous fs-operations on start up. Worst the amount of sync-operations doesn't seem to have an upper bound but perf depends on the number and size of dirty files. All other startup-operations, like establishing IPC channels or setting up telemetry etc, have to wait for this which is bad. Instead, the backup-service should perform async fs operations, and waiting should happen as late as possible, e.g. *before* showing the window but *after* running other startup-tasks.

**Comment by jrieken** [comment]
related to #15455

**Comment by bpasero** [comment]
We can probably easily convert them to be async and just let the window opening wait on it.

@jrieken I found many more sync calls in renderer and extension host and opened https://github.com/Microsoft/vscode/issues/16382

**Comment by Tyriar** [comment]
Moving to January as endgame has come.

**Comment by jrieken** [comment]
fyi - this 10% of the time we spend in the main side *and* it blocks the opening of the window

<img width="540" alt="screenshot 2018-09-25 at 16 38 10" src="https://user-images.githubusercontent.com/1794099/46021868-d3e56180-c0e1-11e8-8d96-d51ded19a724.png">

**Comment by jrieken** [comment]
@bpasero You could work around this by not using [`backupMainService.isHotExitEnabled`](https://github.com/Microsoft/vscode/blob/515c90a9fe1d8a28c0794420c3d82f7060bce73c/src/vs/code/electron-main/window.ts#L514), but by simply using the configuration service.

**Comment by bpasero** [comment]
@jrieken I don't actually think that the window needs to wait on that particular method but there will be more things the window needs to know when opening, mainly the backup location. 

However, what I do not understand is the fs.writeFile on startup. It looks like we always write the backup configuration file to disk even if nothing has changed? @Tyriar could you check?

All in all, that is what will take most time probably on startup.

**Comment by Tyriar** [comment]
We can probably remove the `saveSync` safely based on the comment as anything reading it will get the paths validated.

**Comment by bpasero** [comment]
I pushed a [change](https://github.com/Microsoft/vscode/commit/8094b7708ac47634d5fd56bfa94aa11eb9cd171c) to init the backup service async on startup. It looks like the `saveSync` method was added because it was removed from some of the methods that were called. In other words, we trigger a save because as part of the startup the backup file contents change. 

As a follow up step, the backup service should not save sync at all but rather coordinated on shutdown, similar to what the storage service does by listening to the `onWillShutdown` event on the main side which allows to join with a promise:

https://github.com/Microsoft/vscode/blob/8abe73932c7733f9e17be38391b50e8565cb4b26/src/vs/code/electron-main/app.ts#L465
